hal/spi_api.h

```diff
@@ -19,220 +18,331 @@
+ * @param[in,out] obj       This.
+ * @param[in,out] args      A pointer to a spi_transfer_args_t object.
+ * @param[in]     context   A context to be used by the callback.
+ * @param[in]     cb        A callback invoked upon completion of this transaction.
  *
- * Blocks until the SPI peripheral can be written to
- * @param[in] obj   The SPI peripheral to write
- * @param[in] value The value to write
+ * @return A reference counted handle to the transaction.
  */
-void spi_slave_write(spi_t *obj, int value);
+spi_async_handle_t* spi_async_transfer(
+    spi_t *obj,
+    spi_transfer_args_t *args,
+    void *context,
+    spi_event_f cb);
 
-/** Checks if the specified SPI peripheral is in use
+/**
+ * Aborts the transaction referenced by the spi_async_handle_t.
+ *
+ * This function notifies the lowlevel implementation that the given transaction must be cancelled
+ * (if not already started) or aborted (if currently running) and returns.
+ *
+ * The callback associated with this spi_async_handle_t will be invoked with SPI_EVENT_TYPE_ON_ABORT
+ * as its evtype argument.
  *
- * @param[in] obj The SPI peripheral to check
- * @return non-zero if the peripheral is currently transmitting
+ * This does nothing when used on an already cancelled/aborted transaction.
+ *
+ * @param[in] handle This.
+ *
+ * The handle is consumed in this operation and should no longer be used.
  */
-int  spi_busy(spi_t *obj);
+void spi_async_abort(spi_async_handle_t* handle);
 
-/** Get the module number
+/**
+ * Tells the low-level driver that the upper layer is no longer keeping this handle.
+ *
+ * This does nothing if passed a NULL pointer.
  *
- * @param[in] obj The SPI peripheral to check
- * @return The module number
+ * @param[in] handle This.
  */
-uint8_t spi_get_module(spi_t *obj);
+void spi_async_free_handle(spi_async_handle_t* handle);
 
-/**@}*/
-
-#if DEVICE_SPI_ASYNCH
 /**
- * \defgroup AsynchSPI Asynchronous SPI Hardware Abstraction Layer
- * @{
+ * Frees the SPI instance.
+ *
+ * A SPI instance cannot be released if any reference to an asynchronous transaction is still alive.
+ *
+ * @warning An asynchronous transaction might be completed/errored/aborted and still alive if
+ * `spi_async_free_handle()` has not beed called yet on it.
+ *
+ * @param[in,out] obj   This.
+ *
+ * @return `true` on success.
  */
+bool spi_free(spi_t *obj);
 
-/** Begin the SPI transfer. Buffer pointers and lengths are specified in tx_buff and rx_buff
- *
- * @param[in] obj       The SPI object that holds the transfer information
- * @param[in] tx        The transmit buffer
- * @param[in] tx_length The number of bytes to transmit
- * @param[in] rx        The receive buffer
- * @param[in] rx_length The number of bytes to receive
- * @param[in] bit_width The bit width of buffer words
- * @param[in] event     The logical OR of events to be registered
- * @param[in] handler   SPI interrupt handler
- * @param[in] hint      A suggestion for how to use DMA with this transfer
+/*
```

Comment:
I don't get why do we have it, can't we reimplement the slave driver using the above? If not let's create a doxygen group spi_hal_deprecated or something like that.

Reply:
we could implement spi_read and spi_write but not spi_data_available.
However, we would not need it by handling externally the SS event and triggering a spi_tranfer from there instead of polling on the peripheral to see if something is coming in.